lib/paragraph_maker.rb

```diff
@@ -0,0 +1,260 @@
+        node.replace("<split>")
+      else
+        # Replace the <br> sequence with a special tag (to be processed later)
+        # that indicates that whatever paragraph we're in needs to be split at
+        # this point, and there needs to be a <p>&nbsp;</p> paragraph inserted
+        # into the split. This corresponds to having 3+ <br> tags in a row.
+        node.replace("<split long>")
+      end
+    end
+  end
+
+  # Go through all nodes that should be inside paragraph tags, and make sure
+  # they're wrapped:
+  def wrap_all(root)
+    # We don't use traverse(root) here to handle the traversal because we're
+    # doing something much more complex than usual with our children.
+    return if TAG_NAMES_TO_SKIP.include?(root.name) || in_paragraph?(root) || root.name == "p"
+
+    # Divide up the list of children into sublists based on whether the node
+    # needs to have a paragraph wrapped around it, or not.
+    chunks = root.children.chunk do |node|
+      node.text? || node.cdata? || TAG_NAMES_TO_WRAP.include?(node.name)
+    end
+
+    chunks.each do |should_wrap, children|
+      if should_wrap
+        # Create a paragraph object, and then transfer all of the children in
+        # this chunk to it.
+        paragraph = root.document.create_element("p")
+        children.first.add_previous_sibling(paragraph)
+        children.each do |node|
+          paragraph.add_child(node)
+        end
+      else
+        # These nodes don't need to be wrapped in a paragraph node, but some of
+        # their children might, so we have to process them.
+        children.each do |node|
+          wrap_all(node)
+        end
+      end
+    end
+  end
+
+  # Given a node, "split" its parent at that node and move the node upwards one
+  # level -- that is, split its siblings by whether they're left/right of the
+  # node, and split the parent into two copies, one containing the left siblings
+  # and one containing the right siblings.
+  #
+  # For example, if we have something like this:
+  #   <ul class="actions">
+  #     <li>First</li>
+  #     <li>Second</li>
+  #     <li>Third</li>
+  #   </ul>
+  # Then calling split_parent on the second list item will produce this:
+  #   <ul class="actions">
+  #     <li>First</li>
+  #   </ul>
+  #   <li>Second</li>
+  #   <ul class="actions">
+  #     <li>Third</li>
+  #   </ul>
+  def split_parent(node)
+    # The easy cases: if this node is the first or last node in its parent, then
+    # there's no point in a proper split, because one of the two tags would just
+    # be empty. Instead, we rearrange ourselves relative to our parent.
+    return node.parent.add_previous_sibling(node) if node.previous_sibling.nil?
+    return node.parent.add_next_sibling(node) if node.next_sibling.nil?
+
+    # Create a shallow copy of the parent:
+    new_parent = node.parent.dup(0)
+
+    # Copy over all of the attributes:
+    node.parent.attributes.each do |name, attribute|
```

Comment:
Should we have a spec for keeping the attributes on the new split parent?

Reply:
There's this pre-existing spec: https://github.com/otwcode/otwarchive/blob/d74d404f052b4e60e223c712fff619844695806d/spec/lib/html_cleaner_spec.rb#L961-L966
Do you want me to add more?

Reply:
Oh this works! No need to add more, I just didn't find it.